### PR TITLE
implemented low-level replace_all function

### PR DESCRIPTION
**Pull request by Aman-Godara**
#366 

Current implement provides with `replace_overlapping` optional dummy argument which is `logical`.
This argument helps provide functionality of both OPTION: 1 and OPTION: 2 of [overlapping problem](https://github.com/fortran-lang/stdlib/issues/366#issuecomment-808897041) depending upon whether it is set to `.true.` or `.false.`.

[EDIT]: `replace_overlapping` has been removed from `replace_all` now.

Tasks:

- [x] Implemented `replace_all` function
- [x] Added some test cases for `replace_all` to show the behaviour of the function
- [x] Added more test cases for `replace_all`
- [x] Document the function
- [x] Pure or Elemental -> Pure.

### CODE REVIEW

**Comment by Aman-Godara** [comment]
Although I have added `replace_overlapping` feature to the function but I would like to know the opinion of the community on what behaviour do they find more intuitive from the function in case of overlapping substrings `pattern`.

**Comment by Aman-Godara** [comment]
This comment is regarding commit e325c2b.
replace_all(`string`, `pattern`, `replacement`, `replace_overlapping`)

Please note that for input: `replace_all("abcc", "abc", "ab")` 
commit e325c2b returns `"abc"` even if `replace_overlapping` is set to `.true.` because we are returning a new string as the output. Function is using the input string just to read from it and not to write in it.

***
If we would have designed `replace_all` function differently (i.e. by keeping `replace_all` as subroutine and making few other modifications in the algorithm), function could have been made to give `"ab"` as the output. Answer `"ab"` occurs as such:
**1st replacement**: `"abcc"` -> `"abc"`      (replaced `"abc"` with `"ab"`)
**2nd replacement**: `"abc"` -> `"ab"`        (replaced `"abc"` with `"ab"`)
**No further replacements.**

Both these replacements are happening in the one single call of `replace_all`. Here after **1st replacement**, `"ab"` fitted very well with `"c"` acting as the `pattern` for the **2nd replacement**.

Moreover, this behaviour cannot be achieved by running `replace_all` (as implement in commit e325c2b even with `replace_overlapping` as `.true.`) several times, as:
running `replace_all` 2 times on `"acaaca"` to replace **pattern**: `"aca"` with **replacement**: `"ac"` returns `"acc"` whereas correct output should be `"acca"` as such:
**1st replacement**: `"acaaca"` -> `"acaca"`
**2nd replacement**: `"acaca"` -> `"acca"`
**No further replacements.**

@arjenmarkus too has highlighted a similar behaviour through this sentence in [his comment](https://github.com/fortran-lang/stdlib/issues/366#issuecomment-810893962):
> If the replacing string is qwqwq you would get an infinitely long string.


Whole aim of the comment was to highlight the problems with subroutine kind of `replace_all` function and to bring forward what implementation e325c2b cannot do.

File: src/stdlib_strings.f90 [hunk 1]
```diff
@@ -499,5 +513,140 @@ pure function compute_lps(string) result(lps_array)
...
+                    p_i = p_i + 1
+                else if (p_i > 1) then
+                    p_i = lps_array(p_i - 1) + 1
+                else
+                    s_i = s_i + 1
+                end if
+            end do
+        end if
+        
+        res = res // slice(string, first=last)
```
Comment (Aman-Godara) [inline src/stdlib_strings.f90:648]: line 648 is equivalent to: `res = res // string(last : length_string)`

File: src/stdlib_strings.f90 [hunk 2]
```diff
@@ -499,5 +513,140 @@ pure function compute_lps(string) result(lps_array)
...
+    !> Replaces all occurrences of substring 'pattern' in the input 'string'
+    !> with the replacement 'replacement'
+    !> Returns a new string
+    pure function replace_all_string_string_string(string, pattern, replacement) result(res)
+        type(string_type), intent(in) :: string
+        type(string_type), intent(in) :: pattern
+        type(string_type), intent(in) :: replacement
+        type(string_type) :: res
+
+        res = string_type(replace_all_char_char_char(char(string), &
```
Comment (Aman-Godara) [inline src/stdlib_strings.f90:525]: Should I call `replace_all` instead of `replace_all_char_char_char` in line 525?

Comment (awvwgk) [reply src/stdlib_strings.f90:525]: It's preferable to call the generic interface here

**Review by awvwgk — APPROVED** [review]
Looks good to me.

**Comment by Aman-Godara** [comment]
On page 124 of https://j3-fortran.org/doc/year/18/18-007r1.pdf, It is mentioned that output of character slicing is of length `max(l - f + 1, 0)` which is exactly what is desired in here by the function `replace_all`. Thus, I am making use of intrinsic character slicing instead of `slice` in line 633 and 648. This will slightly improve the performance of the function.

image

**Comment by awvwgk** [comment]
@fortran-lang/push-access Can anyone help out with reviewing this patch to allow @Aman-Godara to move forward with his GSoC project?

**Review by milancurcic — APPROVED** [review]
Thank you @Aman-Godara, I think this is a very nice addition.

I can't find the explanation in this thread or in the related issue: Why did you choose `pure` over `elemental`?

**Comment by milancurcic** [comment]
> I can't find the explanation in this thread or in the related issue: Why did you choose pure over elemental?

All good, I see why, please disregard :+1:

**Comment by Aman-Godara** [comment]
> I can't find the explanation in this thread or in the related issue: Why did you choose `pure` over `elemental`?

Let me still mention it for future reference,
if the length of `replacement` string is not equal to the length of `pattern` string, then different outputs might have varying length which become problematic to be stored in a single output array.

for example:
`replace_all (string, pattern, replacement)`

`replace_all(["abcabcT", "bcdabcT"], ["abc", "bcd"], ["X", "Y"])`
output array in this case will be: `["XXT", "YabcT"]` which cannot be stored in a single output array because of varying lengths of different outputs `"XXT"` and `"YabcT`.

**Pull request merged**